iree/samples/variables_and_state/variables_and_state.ipynb

```diff
@@ -0,0 +1,471 @@
+{
+  "nbformat": 4,
+  "nbformat_minor": 0,
+  "metadata": {
+    "colab": {
+      "name": "variables_and_state.ipynb",
+      "provenance": [],
+      "collapsed_sections": []
+    },
+    "kernelspec": {
+      "display_name": "Python 3",
+      "name": "python3"
+    }
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "j_IgB6_3npkE"
+      },
+      "source": [
+        "```\n",
+        "Copyright 2021 The IREE Authors\n",
+        " \n",
+        "Licensed under the Apache License v2.0 with LLVM Exceptions.\n",
+        "See https://llvm.org/LICENSE.txt for license information.\n",
+        "SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception\n",
+        "```"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "h5s6ncerSpc5"
+      },
+      "source": [
+        "# Variables and State\n",
+        "\n",
+        "This notebook\n",
+        "\n",
+        "1. Creates a TensorFlow program with basic tf.Variable use\n",
+        "2. Imports that program into IREE's compiler\n",
+        "3. Compiles the imported program to an IREE VM bytecode module\n",
+        "4. Tests running the compiled VM module using IREE's runtime\n",
+        "5. Downloads compilation artifacts for use with the native (C API) sample application"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "hHCmr6iGjAJN"
+      },
+      "source": [
+        "## Setup"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "k2SdyC52_QYa"
+      },
+      "source": [
+        "%%capture\n",
+        "!python -m pip install iree-compiler-snapshot iree-runtime-snapshot iree-tools-tf-snapshot -f https://github.com/google/iree/releases"
+      ],
+      "execution_count": 9,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "s2bScbYkP6VZ",
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "outputId": "16907053-5afe-48de-bf7f-48cf1ea7a63a"
+      },
+      "source": [
+        "#@title Imports and setup\n",
+        "\n",
+        "import os\n",
+        "import tempfile\n",
+        "import tensorflow as tf\n",
+        "from IPython.display import clear_output\n",
+        "\n",
+        "from iree import runtime as ireert\n",
+        "from iree.tf.support import module_utils\n",
+        "from iree.compiler import compile_str\n",
+        "from iree.compiler import tf as tfc\n",
+        "\n",
+        "ARTIFACTS_DIR = os.path.join(tempfile.gettempdir(), \"iree\", \"colab_artifacts\")\n",
+        "os.makedirs(ARTIFACTS_DIR, exist_ok=True)\n",
+        "print(f\"Using artifacts directory '{ARTIFACTS_DIR}'\")"
```

Comment:
This would be a nice thing to condense before duplicating - it'd be nice if the imports were closer to their logical use so that people don't assume they always need them. It'll make it easier to clone the notebook too as you can clone it and then if you don't need runtime or don't import from tf you can kill those cells and not have leftover stuff.

For example, compile_str can move to "Compile the imported MLIR", ireert can move to the "test running the compiled...", tf to the "define a simple counter", and tfc to "import the tensorflow program".

That makes this cell just a notebook-specific setup with clear_output and the tempdir stuff - also easy to copy/duplicate in isolation.

Reply:
This could extend to the pip line too with the runtime snapshot installed before the runtime - then you can delete all cells below the runtime portion and not even pull in that dep, or copy all cells below the compiler portion to run things. The cost of an additional pip install invocation is pretty small - people don't use colab notebooks for production pipelines so it's fine if it takes a second or two more to fully run :)

Reply:
I've been following the general python style of including imports at the top of the file: https://google.github.io/styleguide/pyguide.html#313-imports-formatting. In Colab notebooks it is also nice having your imports resolved up front so you can insert cells anywhere and not worry about shuffling imports around for new uses.

I do like your suggestion though - that would help isolate each section.

Reply:
yeah at least having a big `<hr>` between compiler/runtime seems nice